File: wrappers/php/clkmgr_test.php
#!/usr/bin/php
<?php
/* SPDX-License-Identifier: BSD-3-Clause
 * SPDX-FileCopyrightText: Copyright © 2024 Intel Corporation.
 *
 * Test PHP Clock Manager client
 *
 * @author Christopher Hall <christopher.s.hall@@intel.com>
 * @copyright © 2024 Intel Corporation.
 *
 * @note This is a sample code, not a product! You should use it as a reference.
 *
 * To run in development environment:
 * LD_PRELOAD=../../.libs/libclkmgr.so PHPRC=. ./clkmgr_test.php
 *
 * Run ./php_ini.sh to create php.ini
 *
 */

$signal_flag = false;
$event2Sub = EventOffsetInRange | EventSyncedWithGm | EventAsCapable | EventGmChanged;
$composite_event = EventOffsetInRange | EventSyncedWithGm | EventAsCapable;
$chrony_event = EventOffsetInRange;
$clockSyncData = null;

declare(ticks = 1);

function signal_handler(int $signo, $siginfo): void
{
    global $signal_flag;
    echo ' Exit ...' . PHP_EOL;
    $signal_flag = true;
}

function isPositiveValue(string $optarg, string $errorMessage): int
{
    $ret = intval($optarg);
    if($ret >= 0)
        return $ret;
    fwrite(STDERR, $errorMessage . PHP_EOL);
    exit(2);
}

function isValidTimeout(string $optarg, string $errorMessage): int
{
    $ret = intval($optarg);
    if($ret >= -1)
        return $ret;
    fwrite(STDERR, $errorMessage . PHP_EOL);
    exit(2);
}

function clockManagerGetTime(): void
{
    $ts = new timespec();
    if(ClockManager::getTime($ts)) {
        printf('[clkmgr] Current Time of CLOCK_REALTIME: %ld ns' . PHP_EOL,
                ($ts->tv_sec * 1000000000) + $ts->tv_nsec);
    } else {
        echo 'clock_gettime failed: ' . posix_strerror(posix_get_last_error()). PHP_EOL;
    }
}

function getMonotonicTime()
{
    return microtime(true);
}

function printOut(): void
{
    global $clockSyncData, $event2Sub, $composite_event, $chrony_event;
    $hd2l = '|'.str_repeat('-', 30).'|'.str_repeat('-', 28).'|';
    $hd3b = '| %-28s | %-12d | %-11d |';
    $hd3 = '|'.str_repeat('-', 30).'|'.str_repeat('-', 14).'|'.str_repeat('-', 13).'|';
    clockManagerGetTime();
    printf($hd3 . PHP_EOL . '| %-28s | %-12s | %-11s |' . PHP_EOL, 'Events', 'Event Status', 'Event Count');
    $ptpClock = $clockSyncData->getPtp();
    $sysClock = $clockSyncData->getSysClock();
    if($composite_event != 0) {
        echo $hd3 . PHP_EOL;
        printf($hd3b . PHP_EOL, 'ptp_isCompositeEventMet', $ptpClock->isCompositeEventMet(), $ptpClock->getCompositeEventCount());
        if($composite_event & EventOffsetInRange)
            printf('| - %-26s | %-12s | %-11s |' . PHP_EOL, 'isOffsetInRange', '', '');
        if($composite_event & EventSyncedWithGm)
            printf('| - %-26s | %-12s | %-11s |' . PHP_EOL, 'isSyncedWithGm', '', '');
        if($composite_event & EventAsCapable)
            printf('| - %-26s | %-12s | %-11s |' . PHP_EOL, 'isAsCapable', '', '');
    }
    if($event2Sub != 0) {
        echo $hd3 . PHP_EOL;
        if($event2Sub & EventOffsetInRange)
            printf($hd3b . PHP_EOL, 'ptp_isOffsetInRange', $ptpClock->isOffsetInRange(), $ptpClock->getOffsetInRangeEventCount());
        if($event2Sub & EventSyncedWithGm)
            printf($hd3b . PHP_EOL, 'ptp_isSyncedWithGm', $ptpClock->isSyncedWithGm(), $ptpClock->getSyncedWithGmEventCount());
        if($event2Sub & EventAsCapable)
            printf($hd3b . PHP_EOL, 'ptp_isAsCapable', $ptpClock->isAsCapable(), $ptpClock->getAsCapableEventCount());
        if($event2Sub & EventGmChanged)
            printf($hd3b . PHP_EOL, 'ptp_isGmChanged', $ptpClock->isGmChanged(), $ptpClock->getGmChangedEventCount());
    }
    echo $hd3 . PHP_EOL;
    $gmClockUUID = $ptpClock->getGmIdentityStr();
    printf('| %-28s |     %-19ld ns |' . PHP_EOL , 'ptp_clockOffset', $ptpClock->getClockOffset());
    printf('| %-28s |     %s     |' . PHP_EOL, 'ptp_gmIdentity', $gmClockUUID);
    printf('| %-28s |     %-19ld us |' . PHP_EOL .
           '| %-28s |     %-19ld ns |' . PHP_EOL
           , 'ptp_syncInterval', $ptpClock->getSyncInterval()
           , 'ptp_notificationTimestamp', $ptpClock->getNotificationTimestamp());
    echo $hd2l . PHP_EOL;
    if($clockSyncData->haveSys() != 0) {
        $identityString = '';
        $gmIdentity = $sysClock->getGmIdentity();
        for ($i = 0; $i < 4; $i++) {
            $byte = ($gmIdentity >> (8 * (3 - $i))) & 0xFF;
            if($byte == 0 || $byte == 9) {
                $identityString .= ' ';
            } else {
                $s = chr($byte);
                if(preg_match('/[[:print:]]/', $s)) {
                    $identityString .= $s;
                } else {
                    $identityString .= '.';
                }
            }
        }
        printf($hd3b . PHP_EOL .
            $hd2l . PHP_EOL .
            '| %-28s |     %-19ld ns |' . PHP_EOL .
            '| %-28s |     %-19s    |' . PHP_EOL .
            '| %-28s |     %-19ld us |' . PHP_EOL .
            '| %-28s |     %-19ld ns |' . PHP_EOL .
            $hd2l . PHP_EOL . PHP_EOL
            , 'chrony_isOffsetInRange', $sysClock->isOffsetInRange(),
                $sysClock->getOffsetInRangeEventCount()
            , 'chrony_clockOffset', $sysClock->getClockOffset()
            , 'chrony_gmIdentity', $identityString
            , 'chrony_syncInterval', $sysClock->getSyncInterval()
            , 'chrony_notificationTimestamp', $sysClock->getNotificationTimestamp());
    }
}

function main(): void
{
    global $signal_flag, $argv, $event2Sub, $composite_event, $chrony_event, $clockSyncData;
    $ptp4lClockOffsetThreshold = 100000;
    $chronyClockOffsetThreshold = 100000;
    $subscribeAll = false;
    $userInput = false;
    $idleTime = 1;
    $timeout = 10;
    $index = []; # Array of indexes to subscribe
    $clockSyncData = new ClockSyncData();
    $ptp4lSub = new PTPClockSubscription();
    $chronySub = new SysClockSubscription();
    $overallSub = []; # Array of ClockSyncSubscription
    $options = getopt('aps:c:n:l:i:t:m:h');
    if(array_key_exists('h', $options)) {
        $event2SubHex = dechex($event2Sub);
        $composite_eventHex = dechex($composite_event);
        $chrony_eventHex = dechex($chrony_event);
        $me = basename($argv[0]);
        echo <<< END
        Usage of $me :
        Options:
          -a subscribe to all time base indices
             Default: timeBaseIndex: 1
          -p enable user to subscribe to specific time base indices
          -s subscribe_event_mask
             Default: 0x$event2SubHex
             Bit 0: EventOffsetInRange
             Bit 1: EventSyncedWithGm
             Bit 2: EventAsCapable
             Bit 3: EventGmChanged
          -c composite_event_mask
             Default: 0x$composite_eventHex
             Bit 0: EventOffsetInRange
             Bit 1: EventSyncedWithGm
             Bit 2: EventAsCapable
          -n chrony_event_mask
             Default: 0x$chrony_eventHex
             Bit 0: EventOffsetInRange
          -l gm offset threshold (ns)
             Default: $ptp4lClockOffsetThreshold ns
          -i idle time (s)
             Default: $idleTime s
          -m chrony offset threshold (ns)
             Default: $chronyClockOffsetThreshold ns
          -t timeout in waiting notification event (s)
             Default: $timeout s
             -1 : wait indefinitely until at least an event change occurs
              0 : retrieve the latest clock sync data immediately
             >0 : wait up to the specified number of seconds for an event

        END;
        return;
    }
    if(array_key_exists('a', $options))
        $subscribeAll = true;
    else if(array_key_exists('p', $options))
        $userInput = true;
    if(array_key_exists('s', $options))
        $event2Sub = intval($options['s'], 0);
    if(array_key_exists('c', $options))
        $composite_event = intval($options['c'], 0);

    if(array_key_exists('l', $options))
        $ptp4lClockOffsetThreshold = isPositiveValue($options['l'],
            'Invalid ptp4l GM Offset threshold!');
    if(array_key_exists('i', $options))
        $idleTime = isPositiveValue($options['i'], 'Invalid idle time!');
    if(array_key_exists('t', $options))
        $timeout = isValidTimeout($options['t'], 'Invalid timeout!');
    if(array_key_exists('m', $options))
        $chronyClockOffsetThreshold = isPositiveValue($options['m'],
            'Invalid Chrony Offset threshold!');
    if(array_key_exists('n', $options))
        $chrony_event = intval($options['n'], 0);
    $ptp4lSub->setEventMask($event2Sub);
    $ptp4lSub->setClockOffsetThreshold($ptp4lClockOffsetThreshold);
    $ptp4lSub->setCompositeEventMask($composite_event);
    $chronySub->setEventMask($chrony_event);
    $chronySub->setClockOffsetThreshold($chronyClockOffsetThreshold);
    echo '[clkmgr] set subscribe event : 0x' .
        dechex($ptp4lSub->getEventMask()) . PHP_EOL;
    echo '[clkmgr] set composite event : 0x' .
        dechex($ptp4lSub->getCompositeEventMask()) . PHP_EOL;
    echo "GM Offset threshold: $ptp4lClockOffsetThreshold ns" . PHP_EOL;
    echo '[clkmgr] set chrony event : 0x' .
        dechex($chronySub->getEventMask()) . PHP_EOL;
    echo "Chrony Offset threshold: $chronyClockOffsetThreshold ns" . PHP_EOL;

    if($userInput) {
        $line = readline('Enter the time base indices to subscribe ' .
              '(comma-separated, default is 1): ');
        if(strlen($line) > 0) {
            foreach(explode(',', $line) as $str) {
                $idx = intval($str);
                if($idx <= 0) {
                    fwrite(STDERR, "Invalid time base index: $line!" . PHP_EOL);
                    exit(2);
                }
                $index[] = $idx;
            }
        } else {
            echo 'Invalid input. Using default time base index 1.' . PHP_EOL;
        }
    }
    pcntl_signal(SIGINT, 'signal_handler');
    pcntl_signal(SIGTERM, 'signal_handler');
    pcntl_signal(SIGHUP, 'signal_handler');

    $dieMsg = '';
    if(!ClockManager::connect()) {
        $dieMsg = '[clkmgr] failure in connecting !!!';
        goto do_exit;
    }
    sleep(1);
    echo '[clkmgr] List of available clock:' . PHP_EOL;

    # Print out each member of the Time Base configuration
    $size = TimeBaseConfigurations::size();
    for($i = 1; $i <= $size; $i++) {
        $cfg = TimeBaseConfigurations::getRecord($i);
        $idx = $cfg->index();
        echo "TimeBaseIndex: $idx" . PHP_EOL;
        echo 'timeBaseName: ' . $cfg->name() . PHP_EOL;
        $Subscribe = new ClockSyncSubscription();
        if($cfg->havePtp()) {
            $ptp = $cfg->ptp();
            echo 'interfaceName: ' . $ptp->ifName() . PHP_EOL;
            echo 'transportSpecific: ' . $ptp->transportSpecific() . PHP_EOL;
            echo 'domainNumber: ' . $ptp->domainNumber() . PHP_EOL;
            $Subscribe->setPtpSubscription($ptp4lSub);
        }
        if($cfg->haveSysClock())
            $Subscribe->setSysSubscription($chronySub);
        $overallSub[$idx] = $Subscribe;
        if($subscribeAll)
            $index[] = $idx;
    }

    # Default
    if(count($index) == 0)
        $index[] = 1;

    foreach($index as $idx) {
        if(!TimeBaseConfigurations::isTimeBaseIndexPresent($idx)) {
            $dieMsg = "[clkmgr] Index $idx does not exist";
            goto do_exit;
        }
        echo "[clkmgr] Subscribe to time base index: $idx" . PHP_EOL;
        if(!ClockManager::subscribe($overallSub[$idx], $idx, $clockSyncData)) {
            $dieMsg = '[clkmgr] Failure in subscribing to clkmgr Proxy !!!';
            goto do_exit;
        }
        printf('[clkmgr][%.3f] Obtained data from Subscription Event:' . PHP_EOL,
            getMonotonicTime());
        printOut();
    }
    sleep(1);

    $hd2l = '|'.str_repeat('-', 27).'|'.str_repeat('-', 28).'|';
    $hd3b = '| %-25s | %-12d | %-11d |';
    $hd3 = '|'.str_repeat('-', 27).'|'.str_repeat('-', 14).
        '|'.str_repeat('-', 13).'|';

    while(!$signal_flag) {
        next_loop:
        foreach($index as $idx) {
            if($signal_flag)
                goto do_exit;
            printf('[clkmgr][%.3f] Waiting Notification from ' .
                "time base index $idx ..." . PHP_EOL, getMonotonicTime());
            $retval = ClockManager::statusWait($timeout, $idx, $clockSyncData);
            switch($retval) {
                case SWRLostConnection:
                    printf('[clkmgr][%.3f] Terminating: ' .
                        'lost connection to clkmgr Proxy' . PHP_EOL, getMonotonicTime());
                    goto do_exit;
                case SWRInvalidArgument:
                    $dieMsg = sprintf('[clkmgr][%.3f] Terminating: Invalid argument' . PHP_EOL,
                        getMonotonicTime());
                    goto do_exit;
                case SWRNoEventDetected:
                    printf('[clkmgr][%.3f] No event status changes identified ' .
                        "in $timeout seconds." . PHP_EOL . PHP_EOL, getMonotonicTime());
                    printf("[clkmgr][%.3f] sleep for $idleTime seconds..." . PHP_EOL . PHP_EOL,
                        getMonotonicTime());
                    if($signal_flag)
                        goto do_exit;
                    sleep($idleTime);
                    goto next_loop;
                case SWREventDetected:
                    printf('[clkmgr][%.3f] Obtained data from Notification Event:' . PHP_EOL,
                        getMonotonicTime());
                    break;
                default:
                    printf('[clkmgr][%.3f] Warning: Should not enter ' .
                           "this switch case, unexpected status code $retval" . PHP_EOL,
                           getMonotonicTime());
                    goto do_exit;
            }
            printOut();
            printf('[clkmgr][%.3f] sleep for %d seconds...' . PHP_EOL . PHP_EOL,
                getMonotonicTime(), $idleTime);
            if($signal_flag)
                goto do_exit;
            sleep($idleTime);
        }
    }

    do_exit:

    ClockManager::disconnect();

    if(strlen($dieMsg) > 0) {
        fwrite(STDERR, $dieMsg . PHP_EOL);
        exit(2);
    }
}
main();

File: wrappers/perl/clkmgr_test.pl
#!/usr/bin/perl
# SPDX-License-Identifier: BSD-3-Clause
# SPDX-FileCopyrightText: Copyright © 2024 Intel Corporation.
#
# Test Perl Clock Manager client
#
# @author Christopher Hall <christopher.s.hall@@intel.com>
# @copyright © 2024 Intel Corporation.
#
# @note This is a sample code, not a product! You should use it as a reference.
#
# To run in development environment:
# LD_PRELOAD=../../.libs/libclkmgr.so PERL5LIB=. ./clkmgr_test.pl
#
###############################################################################

use threads;
use threads::shared;
use File::Basename;
use POSIX;
use Time::HiRes;
use Getopt::Std;
use ClkMgrLib;

my $signal_flag :shared = 0;
my $clockSyncData = ClkMgrLib::ClockSyncData->new;
my $event2Sub = $ClkMgrLib::EventOffsetInRange | $ClkMgrLib::EventSyncedWithGm |
    $ClkMgrLib::EventAsCapable | $ClkMgrLib::EventGmChanged;
my $composite_event = $ClkMgrLib::EventOffsetInRange |
    $ClkMgrLib::EventSyncedWithGm | $ClkMgrLib::EventAsCapable;
my $chrony_event = $ClkMgrLib::EventOffsetInRange;

sub signal_handler()
{
    my($sig) = @_;
    print " Exit ...\n";
    $signal_flag = 1;
}

sub isPositiveValue
{
    my ($optarg, $errorMessage) = @_;
    my $ret = int($optarg);
    die $errorMessage unless $ret >= 0;
    $ret;
}

sub isValidTimeout
{
    my ($optarg, $errorMessage) = @_;
    my $ret = int($optarg);
    die $errorMessage unless $ret >= -1 ;
    $ret;
}

sub ClockManagerGetTime
{
    my $ts = ClkMgrLib::timespec->new;
    if(ClkMgrLib::ClockManager::getTime($ts)) {
        printf "[clkmgr] Current Time of CLOCK_REALTIME: %ld ns\n",
                ($ts->swig_tv_sec_get * 1000000000) + $ts->swig_tv_nsec_get;
    } else {
        POSIX::perror('clock_gettime failed');
    }
}

sub getMonotonicTime
{
    Time::HiRes::clock_gettime(Time::HiRes::CLOCK_MONOTONIC);
}

sub printOut
{
    ClockManagerGetTime;
    my $hd2l = '|'.('-'x 30).'|'.('-'x 28).'|';
    my $hd3b = '| %-28s | %-12d | %-11d |';
    my $hd3 = '|'.('-'x 30).'|'.('-'x 14).'|'.('-'x 13).'|';
    printf "$hd3\n" .
        "| %-28s | %-12s | %-11s |\n",
        'Events', 'Event Status', 'Event Count';
    print "$hd3\n";
    my $ptpClock = $clockSyncData->getPtp();
    my $sysClock = $clockSyncData->getSysClock();
    if ($clockSyncData->havePTP()) {
        if ($composite_event != 0) {
            printf "$hd3b\n", 'ptp_isCompositeEventMet',
                $ptpClock->isCompositeEventMet(),
                $ptpClock->getCompositeEventCount();
            printf "| - %-26s | %-12s | %-11s |\n", 'isOffsetInRange', '', ''
                if $composite_event & $ClkMgrLib::EventOffsetInRange;
            printf "| - %-26s | %-12s | %-11s |\n",
                'isSyncedWithGm', '', ''
                if $composite_event & $ClkMgrLib::EventSyncedWithGm;
            printf "| - %-26s | %-12s | %-11s |\n", 'isAsCapable', '', ''
                if $composite_event & $ClkMgrLib::EventAsCapable;
        }
        if ($event2Sub != 0) {
            print "$hd3\n";
            printf "$hd3b\n", 'ptp_isOffsetInRange',
                $ptpClock->isOffsetInRange(),
                $ptpClock->getOffsetInRangeEventCount()
                if $event2Sub & $ClkMgrLib::EventOffsetInRange;
            printf "$hd3b\n", 'ptp_isSyncedWithGm',
                $ptpClock->isSyncedWithGm(),
                $ptpClock->getSyncedWithGmEventCount()
                if $event2Sub & $ClkMgrLib::EventSyncedWithGm;
            printf "$hd3b\n", 'ptp_isAsCapable',
                $ptpClock->isAsCapable(), $ptpClock->getAsCapableEventCount()
                if $event2Sub & $ClkMgrLib::EventAsCapable;
            printf "$hd3b\n", 'ptp_isGmChanged',
                $ptpClock->isGmChanged(), $ptpClock->getGmChangedEventCount()
                if $event2Sub & $ClkMgrLib::EventGmChanged;
        }
        print "$hd3\n";

        printf "| %-28s |     %-19ld ns |\n"
            , 'ptp_clockOffset', $ptpClock->getClockOffset();
        my $gmClockUUID = $ptpClock->getGmIdentity();
        my @gm_identity;
        # Copy the uint64_t into the array
        for (my $i = 0; $i < 8; $i++) {
            $gm_identity[$i] = ($gmClockUUID >> (8 * (7 - $i))) & 0xff;
        }
        printf "| %-28s |     %02x%02x%02x.%02x%02x.%02x%02x%02x     |\n",
            'ptp_gmIdentity',
            $gm_identity[0], $gm_identity[1],
            $gm_identity[2], $gm_identity[3],
            $gm_identity[4], $gm_identity[5],
            $gm_identity[6], $gm_identity[7];
        printf "| %-28s |     %-19ld us |\n" .
            "| %-28s |     %-19ld ns |\n" .
            "$hd2l\n"
            , 'ptp_syncInterval', $ptpClock->getSyncInterval()
            , 'ptp_notificationTimestamp', $ptpClock->getNotificationTimestamp();
    }
    if ($clockSyncData->haveSys()) {
        my $identity_string = '';
        for my $i (0..3) {
            my $byte = ($sysClock->getGmIdentity() >> (8 * (3 - $i))) & 0xFF;
            if($byte == 0 || $byte == 9) {
                $identity_string .= ' ';
            } else {
                my $s = chr($byte);
                if($s =~ /[[:print:]]/) {
                    $identity_string .= $s;
                } else {
                    $identity_string .= '.';
                }
            }
        }
        printf "$hd3b\n" .
            "$hd2l\n" .
            "| %-28s |     %-19ld ns |\n" .
            "| %-28s |     %-19s    |\n" .
            "| %-28s |     %-19ld us |\n" .
            "| %-28s |     %-19ld ns |\n" .
            "$hd2l\n\n"
            , 'chrony_isOffsetInRange', $sysClock->isOffsetInRange(),
                $sysClock->getOffsetInRangeEventCount()
            , 'chrony_clockOffset', $sysClock->getClockOffset()
            , 'chrony_gmIdentity', $identity_string
            , 'chrony_syncInterval', $sysClock->getSyncInterval()
            , 'chrony_notificationTimestamp', $ptpClock->getNotificationTimestamp();
    }
}

sub main
{
    my $ptp4lClockOffsetThreshold = 100000;
    my $chronyClockOffsetThreshold = 100000;
    my $subscribeAll = 0;
    my $userInput = 0;
    my $idleTime = 1;
    my $timeout = 10;
    my @index; # Array of indexes to subscribe

    my $ptp4lSub = ClkMgrLib::PTPClockSubscription->new;
    my $chronySub = ClkMgrLib::SysClockSubscription->new;

    my @overallSub; # Array of ClkMgrLib::ClockSyncSubscription

    my $ret = getopts('aps:c:n:l:i:t:m:h');
    if($opt_h || !$ret) {
        my $event2SubHex = sprintf '0x%x', $event2Sub;
        my $composite_eventHex = sprintf '0x%x', $composite_event;
        my $chrony_eventHex = sprintf '0x%x', $chrony_event;
        my $name = basename $0;
        my $help = <<"END_MESSAGE";
Usage of $name :
Options:
  -a subscribe to all time base indices
     Default: timeBaseIndex: 1
  -p enable user to subscribe to specific time base indices
  -s subscribe_event_mask
     Default: $event2SubHex
     Bit 0: EventOffsetInRange
     Bit 1: EventSyncedWithGm
     Bit 2: EventAsCapable
     Bit 3: EventGmChanged
  -c composite_event_mask
     Default: $composite_eventHex
     Bit 0: EventOffsetInRange
     Bit 1: EventSyncedWithGm
     Bit 2: EventAsCapable
  -n chrony_event_mask
     Default: $chrony_eventHex
     Bit 0: EventOffsetInRange
  -l gm offset threshold (ns)
     Default: $ptp4lClockOffsetThreshold ns
  -i idle time (s)
     Default: $idleTime s
  -m chrony offset threshold (ns)
     Default: $chronyClockOffsetThreshold ns
  -t timeout in waiting notification event (s)
     Default: $timeout s
     -1 : wait indefinitely until at least an event change occurs
      0 : retrieve the latest clock sync data immediately
     >0 : wait up to the specified number of seconds for an event
END_MESSAGE
        $ret ? print($help) : die $help;
        return;
    }
    $subscribeAll = $opt_a ? 1 : 0;
    $userInput = $subscribeAll ? 0 : ($opt_p ? 1 : 0);
    $event2Sub = POSIX::strtoul($opt_s, 0) if $opt_s;
    $composite_event = POSIX::strtoul($opt_c, 0) if $opt_c;
    $chrony_event = POSIX::strtoul($opt_n, 0) if $opt_n;
    $ptp4lClockOffsetThreshold = (
        isPositiveValue($opt_l, 'Invalid ptp4l GM Offset threshold!')
    ) if defined $opt_l;
    $idleTime = isPositiveValue($opt_i, 'Invalid idle time!') if defined $opt_i;
    $timeout = isValidTimeout($opt_t, 'Invalid timeout!') if defined $opt_t;
    $chronyClockOffsetThreshold = (
        isPositiveValue($opt_m, 'Invalid Chrony Offset threshold!')
    ) if defined $opt_m;
    $ptp4lSub->setEventMask($event2Sub);
    $ptp4lSub->setClockOffsetThreshold($ptp4lClockOffsetThreshold);
    $ptp4lSub->setCompositeEventMask($composite_event);
    $chronySub->setEventMask($chrony_event);
    $chronySub->setClockOffsetThreshold($chronyClockOffsetThreshold);
    printf "[clkmgr] set subscribe event : 0x%x\n", $ptp4lSub->getEventMask();
    printf "[clkmgr] set composite event : 0x%x\n",
        $ptp4lSub->getCompositeEventMask();
    printf "[clkmgr] set chrony event : 0x%x\n", $chrony_event;
    print "GM Offset threshold: $ptp4lClockOffsetThreshold ns\n";
    print "Chrony Offset threshold: $chronyClockOffsetThreshold ns\n";

    if($userInput) {
        print 'Enter the time base indices to subscribe ' .
              '(comma-separated, default is 1): ';
        my $line = <STDIN>;
        chomp $line;
        unless($line =~ /^$/) {
            for(split /, /,$line) {
                my $idx = int($_);
                die "Invalid time base index: $_!" unless $idx > 0;
                push(@index, $idx);
            }
        } else {
            warn "Invalid input. Using default time base index 1.\n";
        }
    }
    $SIG{'INT'} = \&signal_handler;
    $SIG{'TERM'} = \&signal_handler;
    $SIG{'HUP'} = \&signal_handler;

    my $dieMsg;
    unless(ClkMgrLib::ClockManager::connect()) {
        $dieMsg = '[clkmgr] failure in connecting !!!';
        goto do_exit
    }
    sleep 1;
    print "[clkmgr] List of available clock:\n";
    # Print out each member of the Time Base configuration
    my $size = ClkMgrLib::TimeBaseConfigurations::size();
    for(my $i = 1; $i <= $size; $i++) {
        my $cfg = ClkMgrLib::TimeBaseConfigurations::getRecord($i);
        my $idx = $cfg->index();
        print "TimeBaseIndex: $idx\n";
        print 'timeBaseName: ' . $cfg->name() . "\n";
        my $Subscribe = ClkMgrLib::ClockSyncSubscription->new;
        if($cfg->havePtp()) {
            my $ptp = $cfg->ptp();
            print 'interfaceName: ' . $ptp->ifName() . "\n";
            print 'transportSpecific: ' . $ptp->transportSpecific() . "\n";
            print 'domainNumber: ' . $ptp->domainNumber() . "\n";
            $Subscribe->setPtpSubscription($ptp4lSub);
        }
        $Subscribe->setSysSubscription($chronySub) if $cfg->haveSysClock();
        $overallSub[$idx] = $Subscribe;
        print "\n";
        push(@index, $idx) if $subscribeAll;
    }

    # Default
    push(@index, 1) if 0 == scalar @index;

    for(@index) {
        my $idx = $_;
        if(!ClkMgrLib::TimeBaseConfigurations::isTimeBaseIndexPresent($idx)) {
            $dieMsg = "[clkmgr] Index $idx does not exist";
            goto do_exit;
        }
        print "[clkmgr] Subscribe to time base index: $idx\n";
        unless(ClkMgrLib::ClockManager::subscribe($overallSub[$idx],
            $idx, $clockSyncData)) {
            $dieMsg = '[clkmgr] Failure in subscribing to clkmgr Proxy !!!';
            goto do_exit;
        }
        printf "[clkmgr][%.3f] Obtained data from Subscription Event:\n",
            getMonotonicTime;
        printOut();
    }
    sleep 1;

    while (!$signal_flag) {
        inner_loop: for(@index) {
            goto do_exit if $signal_flag;
            my $idx = $_;
            printf '[clkmgr][%.3f] Waiting Notification ' .
                   "from time base index $idx ...\n", getMonotonicTime;

            my $retval =
                ClkMgrLib::ClockManager::statusWait($timeout, $idx, $clockSyncData);
            if($retval == $ClkMgrLib::SWRLostConnection) {
                printf '[clkmgr][%.3f] Terminating: lost connection to ' .
                    "clkmgr Proxy\n", getMonotonicTime;
                goto do_exit;
            } elsif($retval == $ClkMgrLib::SWRInvalidArgument) {
                $dieMsg = sprintf "[clkmgr][%.3f] Terminating: Invalid argument\n",
                    getMonotonicTime;
                goto do_exit;
            } elsif($retval == $ClkMgrLib::SWRNoEventDetected) {
                printf '[clkmgr][%.3f] No event status changes identified ' .
                    "in %d seconds.\n\n", getMonotonicTime, $timeout;
                printf "[clkmgr][%.3f] sleep for %d seconds...\n\n",
                    getMonotonicTime, $idleTime;
                goto do_exit if $signal_flag;
                sleep $idleTime;
                next inner_loop;
            } elsif($retval == $ClkMgrLib::SWREventDetected) {
                printf "[clkmgr][%.3f] Obtained data from Notification Event:\n",
                    getMonotonicTime;
            } else {
                printf '[clkmgr][%.3f] Warning: Should not enter this switch ' .
                    "case, unexpected status code %d\n", getMonotonicTime, $retval;
                goto do_exit;
            }
            printOut();
            printf "[clkmgr][%.3f] sleep for %d seconds...\n\n",
                getMonotonicTime, idleTime;
            goto do_exit if $signal_flag;
            sleep $idleTime;
        }
    }

do_exit:

    ClkMgrLib::ClockManager::disconnect();
    die $dieMsg if $dieMsg;
}
main;
